src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs

```diff
@@ -185,16 +184,48 @@ public class ParameterBinder
             await modelBinder.BindModelAsync(modelBindingContext);
 
             var modelBindingResult = modelBindingContext.Result;
-            if (modelBindingResult.IsModelSet)
+
+            Validate(
+                actionContext,
+                parameter,
+                metadata,
+                modelBindingContext,
+                modelBindingResult);
+
+            return modelBindingResult;
+        }
+
+        private void Validate(
+            ActionContext actionContext,
+            ParameterDescriptor parameterDescriptor,
+            ModelMetadata modelMetadata,
+            ModelBindingContext bindingContext,
+            ModelBindingResult modelBindingResult)
+        {
+            if (modelBindingResult.IsModelSet || modelMetadata.IsRequired)
             {
-                _validator.Validate(
+                var visitor = new ValidationVisitor(
                     actionContext,
-                    modelBindingContext.ValidationState,
-                    modelBindingContext.ModelName,
-                    modelBindingResult.Model);
+                    _validatorProvider,
+                    _validatorCache,
+                    _modelMetadataProvider,
+                    bindingContext.ValidationState);
+
+                var isValid = visitor.Validate(modelMetadata, bindingContext.ModelName, modelBindingResult.Model, skipNullAtTopLevel: !modelMetadata.IsRequired);
+                if (!isValid)
+                {
+                    // Only add one validation error in the case where both [Required] and
+                    // [BindRequired] are specified
+                    return;
+                }
             }
 
-            return modelBindingResult;
+            if (!modelBindingResult.IsModelSet && modelMetadata.IsBindingRequired)
```

Comment:
Would it be better to move this block out of this method into the calling code? Generally `IsBindingRequired` is handled by the model binding system, and if binding doesn't succeed we don't need to step into the validator

Reply:
That makes sense. In fact having done that the remaining `Validate`-method logic is so minimal I've just inlined it now.

Reply:
Update: Now I've also added back-compatibility to `ParameterBinder` (reintroduced old constructor (marked `[Obsolete]`), which triggers the old behavior), the logic here was getting pretty long-winded so have pulled it back out into a separate method. It's just a `private` method though so this is just a hidden implementation detail we can change any time.